src/collective.jl

```diff
@@ -602,8 +602,10 @@ function Reduce!(rbuf::RBuffer, op::Union{Op,MPI_Op}, root::Integer, comm::Comm)
 end
 
 # Convert user-provided functions to MPI.Op
-Reduce!(rbuf::RBuffer, op, root::Integer, comm::Comm) =
-    Reduce!(rbuf, Op(op, eltype(rbuf)), root, comm)
+function Reduce!(rbuf::RBuffer, op, root::Integer, comm::Comm)
+    op′ = Op(op, eltype(rbuf))
+    GC.@preserve op′ Reduce!(rbuf, op′, root, comm)
```

Comment:
Why is `GC.@preserve` required here?

Reply:
This code creates an operator, but it is only used in the call to `Reduce!`. While `Reduce!` is running, the Julia operator wrapper appears unused to Julia, and the Julia GC can collect it, and its finalizer then frees the respective `MPI_Op`. Since `Reduce!` calls back into Julia that can and does happen, and can free the created `MPI_Op` before `Reduce!` finishes.

These `GC.@preserve` statements prevent segfaults that I observed.

Reply:
I removed the `GC.@preserve` statements as discussed.